# Perdidos en el espacio

Imagina que en el futuro una nave espacial tripulada parte de la Tierra hacia el planeta Neptuno.

Durante el viaje, la tripulación permanece en estado de animación suspendida. En el trayecto la nave
sufre una grave avería y el sistema automático de control la dirige al cuerpo celeste más cercano,
donde realiza un aterrizaje de emergencia. Al posarse sobre la superficie, el sistema de soporte vital
de la nave reanima a la tripulación.

Una vez despiertos, los tripulantes descubren que el sistema de posicionamiento espacial no
funciona y por el momento no tienen comunicación con la Tierra ni pueden mirar al exterior de la
nave por razones de seguridad. Así que no saben dónde están, no tienen ningún dato del exterior,
ni posibilidad de conseguirlo, ni saben cuánto tiempo ha pasado.

En su trayecto entre la Tierra y Neptuno han podido estar cerca de la propia Tierra, Marte, Júpiter,
Saturno, Urano, Neptuno y las lunas de cualquiera de ellos

In [1]:
import random
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from faker import Faker
from sklearn.preprocessing import LabelEncoder

# Datos Experimentales


In [18]:
import random
from faker import Faker

fake = Faker()

planet_g_values = {
    'Tierra': 9.8,
    'Marte': 3.7,
    'Júpiter': 24.8,
    'Saturno': 9.0,
    'Urano': 8.7,
    'Neptuno': 11.0
}

def generate_sample():
    L = round(random.uniform(0.8, 2.5), 2)
    t = round(random.uniform(0.4, 0.7), 2)
    g_calculated = round((2 * L) / (t**2), 2)
    
    g_diffs = {planet: abs(g_calculated - g) for planet, g in planet_g_values.items()}
 
    sorted_planets = sorted(g_diffs.keys(), key=lambda x: g_diffs[x])
    
    top_3_planets = sorted_planets[:3]

    return {
        "ID": fake.uuid4(),
        "Longitud (m)": L,
        "Tiempo (s)": t,
        "g_calculado (m/s²)": g_calculated,
        "Planeta más probable": top_3_planets[0],
        "2do más probable": top_3_planets[1],
        "3ro más probable": top_3_planets[2]
    }

dataset = [generate_sample() for _ in range(500)]

for sample in dataset:
    print(sample)


{'ID': 'fd48bd2b-3c8f-4d0f-82ca-12120fbd6e22', 'Longitud (m)': 2.35, 'Tiempo (s)': 0.45, 'g_calculado (m/s²)': 23.21, 'Planeta más probable': 'Júpiter', '2do más probable': 'Neptuno', '3ro más probable': 'Tierra'}
{'ID': '68efdc37-eaac-4ba9-bf1a-eb08c036f155', 'Longitud (m)': 2.44, 'Tiempo (s)': 0.45, 'g_calculado (m/s²)': 24.1, 'Planeta más probable': 'Júpiter', '2do más probable': 'Neptuno', '3ro más probable': 'Tierra'}
{'ID': '41c7d164-3322-4864-a9ad-44b83ab3fb1e', 'Longitud (m)': 1.09, 'Tiempo (s)': 0.65, 'g_calculado (m/s²)': 5.16, 'Planeta más probable': 'Marte', '2do más probable': 'Urano', '3ro más probable': 'Saturno'}
{'ID': 'eba360f7-76c8-441c-a2cb-669509ab4c84', 'Longitud (m)': 1.48, 'Tiempo (s)': 0.67, 'g_calculado (m/s²)': 6.59, 'Planeta más probable': 'Urano', '2do más probable': 'Saturno', '3ro más probable': 'Marte'}
{'ID': '8f518daf-8298-456c-854a-770b3e53f21f', 'Longitud (m)': 1.93, 'Tiempo (s)': 0.44, 'g_calculado (m/s²)': 19.94, 'Planeta más probable': 'Júpiter', 

In [19]:
csv_filename = "planet_dataset.csv"
df = pd.DataFrame(dataset)
df.to_csv(csv_filename, index=False)

In [21]:
filepath = "planet_dataset.csv"
data_frame = pd.read_csv(filepath)

data_frame.head(500)

ID  Longitud (m)  Tiempo (s)  \
0    fd48bd2b-3c8f-4d0f-82ca-12120fbd6e22          2.35        0.45   
1    68efdc37-eaac-4ba9-bf1a-eb08c036f155          2.44        0.45   
2    41c7d164-3322-4864-a9ad-44b83ab3fb1e          1.09        0.65   
3    eba360f7-76c8-441c-a2cb-669509ab4c84          1.48        0.67   
4    8f518daf-8298-456c-854a-770b3e53f21f          1.93        0.44   
..                                    ...           ...         ...   
495  7f1fdb3a-64c7-4fa7-96b9-e7f4ad493eb8          0.99        0.67   
496  46ddac36-28f9-4be4-a494-da8e31416375          1.94        0.64   
497  4ad27f08-7bfd-446a-b3bc-ddd344e0f148          1.41        0.68   
498  265d0abd-60c0-410e-8b2a-dba574c5931f          2.43        0.45   
499  67776db6-b061-4455-aa8b-8fe82ad7c0d0          2.35        0.59   

     g_calculado (m/s²) Planeta más probable 2do más probable 3ro más probable  
0                 23.21              Júpiter          Neptuno           Tierra  
1                 24.10              Júpiter          Neptuno           Tierra  
2                  5.16                Marte            Urano          Saturno  
3                  6.59                Urano          Saturno            Marte  
4                 19.94              Júpiter          Neptuno           Tierra  
..                  ...                  ...              ...              ...  
495                4.41                Marte            Urano          Saturno  
496                9.47               Tierra          Saturno            Urano  
497                6.10                Marte            Urano          Saturno  
498               24.00              Júpiter          Neptuno           Tierra  
499               13.50              Neptuno           Tierra          Saturno  

[500 rows x 7 columns]

# Fórmula: L = 1/2 x g x t^2

    A.- Con una longitud L= 1.00 m cronometran t = 0.46 s.

    ¿Cuánto vale g y su error?
 
    ¿En qué planeta están, o pueden estar?

In [8]:
import numpy as np

L = 1.00  # m
t = 0.46  # s
delta_L = 0.01  # m
delta_t = 0.01  # s

g = 2 * L / t**2

dg_dL = 2 / t**2
dg_dt = -4 * L / t**3

delta_g = np.sqrt((dg_dL * delta_L)**2 + (dg_dt * delta_t)**2)

g, delta_g

df = pd.DataFrame({'Aceleración de la gravedad medida (g)': [g],
                   'Error en la medición de g': [delta_g]})

print(df)

   Aceleración de la gravedad medida (g)  Error en la medición de g
0                               9.451796                   0.421677


Dado que g = 9.45 m/s^2 + - 0.42 m/s², podemos calcular el rango de valores posibles para g:

Rango inferior: 9.45 - 0.42 = 9.03 m/s^2

Rango superior: 9.45 + 0.42 = 9.87 m/s^2

Teniendo en cuenta los datos de la tabla, podemos saber que la tripulación podría estar en:

Saturno (g = 9.0 m/s^2) ya que 9.0 está dentro del rango de 9.03 a 9.87 m/s^2.

Tierra (g = 9.8 m/s^2) ya que 9.8 también está dentro del rango de 9.03 a 9.87 m/s^2.

    B.- Con una longitud L=2.00 m cronometran t = 0.66 s.

    ¿Cuánto vale g y su error?

    ¿En qué planeta están, o pueden estar?
 
    ¿Con que probabilidad puedes afirmarlo?

In [22]:
L = 2.00  # m
t = 0.66  # s

g = 2 * L / t**2

dg_dL = 2 / t**2
dg_dt = -4 * L / t**3

delta_g = np.sqrt((dg_dL * delta_L)**2 + (dg_dt * delta_t)**2)

g, delta_g

df = pd.DataFrame({'Aceleración de la gravedad medida (g)': [g],
                   'Error en la medición de g': [delta_g]})

print(df)

   Aceleración de la gravedad medida (g)  Error en la medición de g
0                               9.182736                   0.282027


Ahora, podemos calcular el rango de valores posibles para g:

Rango inferior: 9.18 - 0.28 = 8.90 m/s^2
Rango superior: 9.18 + 0.28 = 9.46 m/s^2

Podemos concluir que la tripulación podría estar en:

Saturno (g = 9.0 m/s^2) ya que 9.0 está dentro del rango de 8.90 a 9.46 m/s^2.
Dado que solo Saturno se encuentra dentro del rango calculado, podemos afirmar con cierta confianza que la tripulación está en Saturno en este caso.

# Creación de un Dataset enfocado en Experimentos de Aceleración de Gravedad para determinar la ubicación de una nave espacial en Tierra o Saturno 

Ahora vamos a crear un dataset enfocado en los experimentos realizados en los planetas con mayores porcentajes de posibilidad de estar presentes. Estos experimentos se centran en la Tierra y Saturno, ya que, según los cálculos previos, estos dos planetas mostraron los valores de aceleración de la gravedad más cercanos a los obtenidos en los escenarios analizados.

Este dataset contendrá los datos de las longitudes y tiempos cronometrados de cada experimento, así como los valores de la aceleración de la gravedad calculados y los porcentajes de acierto para cada planeta. Al analizar estos datos, podremos obtener una mejor comprensión de las posibilidades de que la nave espacial se encuentre en la Tierra o en Saturno, y así tomar decisiones más informadas sobre cómo proceder en esta situación.

Al utilizar este enfoque centrado en los planetas con mayores probabilidades, esperamos que podamos obtener resultados más precisos y útiles para determinar la ubicación de la nave espacial y garantizar el éxito de la misión.

In [15]:
import random
fake = Faker()

def generate_sample():
    L = round(random.uniform(0.8, 2.5), 2)
    t = round(random.uniform(0.4, 0.7), 2)
    g_calculated = round((2 * L) / (t**2), 2)
    most_probable_planet = "Tierra" if abs(g_calculated - 9.8) < abs(g_calculated - 9.0) else "Saturno"
    second_most_probable_planet = "Saturno" if most_probable_planet == "Tierra" else "Tierra"
    
    return {
        "ID": fake.uuid4(),
        "Longitud (m)": L,
        "Tiempo (s)": t,
        "g_calculado (m/s²)": g_calculated,
        "Planeta más probable": most_probable_planet,
        "2do más probable": second_most_probable_planet
    }

dataset = [generate_sample() for _ in range(100)]

for sample in dataset:
    print(sample)



{'ID': 'ca8c3e9d-bcf4-4158-9423-3e959d965eee', 'Longitud (m)': 1.7, 'Tiempo (s)': 0.42, 'g_calculado (m/s²)': 19.27, 'Planeta más probable': 'Tierra', '2do más probable': 'Saturno'}
{'ID': '5c360f09-ba41-440d-834b-3f7a0b1f484c', 'Longitud (m)': 1.62, 'Tiempo (s)': 0.61, 'g_calculado (m/s²)': 8.71, 'Planeta más probable': 'Saturno', '2do más probable': 'Tierra'}
{'ID': '65f024bd-96fe-4f0e-8cc5-2c81eeae84b7', 'Longitud (m)': 2.42, 'Tiempo (s)': 0.69, 'g_calculado (m/s²)': 10.17, 'Planeta más probable': 'Tierra', '2do más probable': 'Saturno'}
{'ID': 'cc5fd1e8-697a-4a36-961a-86182aa8d690', 'Longitud (m)': 1.85, 'Tiempo (s)': 0.53, 'g_calculado (m/s²)': 13.17, 'Planeta más probable': 'Tierra', '2do más probable': 'Saturno'}
{'ID': 'bd153d15-a034-4167-99e1-737b179cdf0d', 'Longitud (m)': 1.93, 'Tiempo (s)': 0.51, 'g_calculado (m/s²)': 14.84, 'Planeta más probable': 'Tierra', '2do más probable': 'Saturno'}
{'ID': '6b39d02d-1dcc-49c4-8bc3-4d2283aa8404', 'Longitud (m)': 1.03, 'Tiempo (s)': 0.59

In [16]:
data_frame = pd.DataFrame(dataset)
data_frame.to_csv("generated_dataset.csv", index=False)


In [17]:
filepath = "generated_dataset.csv"
data_frame = pd.read_csv(filepath)

data_frame.head(100)


ID  Longitud (m)  Tiempo (s)  \
0   ca8c3e9d-bcf4-4158-9423-3e959d965eee          1.70        0.42   
1   5c360f09-ba41-440d-834b-3f7a0b1f484c          1.62        0.61   
2   65f024bd-96fe-4f0e-8cc5-2c81eeae84b7          2.42        0.69   
3   cc5fd1e8-697a-4a36-961a-86182aa8d690          1.85        0.53   
4   bd153d15-a034-4167-99e1-737b179cdf0d          1.93        0.51   
..                                   ...           ...         ...   
95  37c557ad-0600-4154-9962-358adc7e9d9a          1.57        0.58   
96  4712ae7b-440a-4440-9225-68aba9685294          1.94        0.57   
97  9cc44df8-4c6c-4a4f-838d-eaec30355c33          1.28        0.44   
98  40aa4adb-83cf-4d2a-9da5-3d81753102b9          1.88        0.55   
99  1c63f2cf-ac29-43f3-9340-614a4e9b9d28          1.16        0.47   

    g_calculado (m/s²) Planeta más probable 2do más probable  
0                19.27               Tierra          Saturno  
1                 8.71              Saturno           Tierra  
2                10.17               Tierra          Saturno  
3                13.17               Tierra          Saturno  
4                14.84               Tierra          Saturno  
..                 ...                  ...              ...  
95                9.33              Saturno           Tierra  
96               11.94               Tierra          Saturno  
97               13.22               Tierra          Saturno  
98               12.43               Tierra          Saturno  
99               10.50               Tierra          Saturno  

[100 rows x 6 columns]

# Análisis Exploratorio de Datos para el experimento de Tierra y Saturno como pruebas para determinar la ubicación de la nave espacial